src/odemis/util/spot.py

```diff
@@ -126,86 +128,260 @@ def SpotIntensity(data, background=None):
-    xc = xc  # + 1 / 2
-    yc = -yc  # + 1 / 2
-    return (xc, yc)
+def FindCenterCoordinates(image, smoothing=True):
+    """
+    Returns the radial symmetry center of the image with sub-pixel resolution.
+    This is the spot center location if there is only a single spot contained
+    in the image.
+
+    Parameters
+    ----------
+    image : array_like
+        The image of which to determine the radial symmetry center.
+    smoothing : boolean
+        Apply a smoothing kernel to the intensity gradient.
+
+    Returns
+    -------
+    pos : tuple
+        Position of the radial symmetry center in px from the center of the
+        image.
+
+    Examples
+    --------
+    >>> img = np.zeros((5, 5))
+    >>> img[2, 2] = 1
+    >>> FindCenterCoordinates(img)
+    (0.0, 0.0)
+
+    """
+    image = np.asarray(image, dtype=np.float64)
+
+    # Compute lattice midpoints (ik, jk).
+    n, m = image.shape
+    jk, ik = np.meshgrid(np.arange(m - 1) + 0.5, np.arange(n - 1) + 0.5)
+
+    # Calculate the intensity gradient.
+    ki = np.array([(1, 1), (-1, -1)])
+    kj = np.array([(1, -1), (1, -1)])
+    dIdi = scipy.signal.convolve2d(image, ki, mode='valid')
+    dIdj = scipy.signal.convolve2d(image, kj, mode='valid')
+    if smoothing:
+        k = np.ones((3, 3)) / 9.
+        dIdi = scipy.signal.convolve2d(dIdi, k, boundary='symm', mode='same')
+        dIdj = scipy.signal.convolve2d(dIdj, k, boundary='symm', mode='same')
+    dI2 = np.square(dIdi) + np.square(dIdj)
+
+    # Discard entries where the intensity gradient magnitude is zero, flatten
+    # the array in the same go.
+    idx = np.flatnonzero(dI2)
+    ik = np.take(ik, idx)
+    jk = np.take(jk, idx)
+    dIdi = np.take(dIdi, idx)
+    dIdj = np.take(dIdj, idx)
+    dI2 = np.take(dI2, idx)
+
+    # Construct the set of equations for a line passing through the midpoint
+    # (ik, jk), parallel to the gradient intensity, in implicit form:
+    # `a*i + b*j + c = 0`, normalized such that `a^2 + b^2 = 1`.
+    dI = np.sqrt(dI2)
+    a = -dIdj / dI
+    b = dIdi / dI
+    c = a * ik + b * jk
+
+    # Weighting: weight by the square of the gradient magnitude and inverse
+    # distance to the centroid of the square of the gradient intensity
+    # magnitude.
+    sdI2 = np.sum(dI2)
+    i0 = np.sum(dI2 * ik) / sdI2
+    j0 = np.sum(dI2 * jk) / sdI2
+    w = np.sqrt(dI2 / np.hypot(ik - i0, jk - j0))
+
+    # Solve the linear set of equations in a least-squares sense.
+    ic, jc = np.linalg.lstsq(np.vstack((w * a, w * b)).T, w * c)[0]
+
+    # Convert from index (top-left) to (center) position information.
+    xc = jc - 0.5 * float(m) + 0.5
+    yc = ic - 0.5 * float(n) + 0.5
+
+    return xc, yc
+
+
+def sedisk(r=3):
+    """
+    Create a flat disk-shaped structuring element with the specified radius r.
+
+    Parameters
+    ----------
+    r : Non-negative integer. Disk radius (default: 3)
+
+    Returns
+    -------
+    nhood : Structuring element
```

Comment:
What is a "structuring element"?! If you don't know either, at least add the type. A 2D numpy array of booleans?